api/src/main/java/jakarta/data/repository/Pagination.java

```diff
@@ -17,55 +17,65 @@
  */
 package jakarta.data.repository;
 
+import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.stream.Collectors;
 import java.util.stream.StreamSupport;
 
-import jakarta.data.repository.KeysetPageable.Cursor;
-
 /**
  * Built-in implementation of Pageable.
  */
-class Pagination implements Pageable {
+final class Pagination implements Pageable {
+
+    private final Optional<Cursor> cursor;
```

Comment:
```suggestion
    private final Cursor cursor;
´´´

https://rules.sonarsource.com/java/RSPEC-3553

Reply:
> https://rules.sonarsource.com/java/RSPEC-3553

The link that you cited about limiting usage of Optional to return types states,

> using Optional on the input side increases the work you have to do in the method without really increasing the value

which is abundantly clear from the changes I needed to make to get the `Cursor` field switched to an `Optional<Cursor>`. Nonetheless, that is the only way of achieving what you asked for in making `cursor()` return `Optional<Cursor>` on an API that we decided will be made compatible with Java records. With records, the constructor-like record declaration needs to accept the same type as the accessor method returns.  `Optional` just doesn't fit well with records. I never wanted to use `Optional` on `Cursor` in the first place. As stated in my earlier comment on the matter, it makes no sense in terms of usage patterns. And as you've seen here, it makes the API more difficult to write, implement, and even use. Switching to `Optional` has nothing to do with whether or not `KeysetPageable` and `Pageable` are made into a single interface, so the issue should never have been combined into this pull in the first place. I will remove the two commits from this pull that switch ever to `Optional` so that the matter can be considered separately.